### PR TITLE
Some simple Coupleable extensions

### LINKED ISSUE #16888: More convenient functions in Coupleable are desired
## Reason
More convenient interfaces function for accessing DoF values and for array variables are desired.

## Design
Extend current available functions as template. I will send in a small PR on the extensions that I need.

## Impact
Make certain derived classes simpler. New functions thus should not have any impact on current capabilities.

### PR DESCRIPTION
**Pull request by YaqiWang**
Close #16888.

### CODE REVIEW

**Comment by YaqiWang** [comment]
This is ready for review.

**Comment by loganharbour** [comment]
The failure is related. It's just hard to see through all of the macro warnings...

```
In file included from /opt/civet/build_1/moose/framework/build/unity_src/variables_Unity.C:9:0:
/opt/civet/build_1/moose/framework/src/variables/MooseVariableFE.C: In member function 'std::__cxx11::string MooseVariableFE< <template-parameter-1-1> >::componentName(unsigned int) const':
/opt/civet/build_1/moose/framework/src/variables/MooseVariableFE.C:58:24: error: 'class MooseVariableBase' has no member named 'arrayVariableComponent'
     return _subproblem.arrayVariableComponent(_var_name, comp);
```

Otherwise, this looks good to me. Glad to see someone else is using `coupledVectorHelper`!

**Comment by YaqiWang** [comment]
Thanks @loganharbour but the error does not make sense. I am calling `Subproblem:: arrayVariableComponent` there why it says `MooseVariableBase` has no such a member?

File: framework/src/variables/MooseVariableFE.C [hunk 1]
```diff
@@ -47,6 +47,19 @@ MooseVariableFE<OutputType>::MooseVariableFE(const InputParameters & parameters)
...
 
+template <typename OutputType>
+std::string
+MooseVariableFE<OutputType>::componentName(const unsigned int comp) const
+{
+  if (comp >= _count)
+    mooseError("Component index must be less than the number of components of variable ",
+               _var_name);
+  if (std::is_same<OutputType, RealEigenVector>::value)
+    return _subproblem.arrayVariableComponent(_var_name, comp);
```
Comment (loganharbour) [inline framework/src/variables/MooseVariableFE.C:58]:
```suggestion
    return this->_subproblem.arrayVariableComponent(_var_name, comp);
```

Try this

Comment (YaqiWang) [reply framework/src/variables/MooseVariableFE.C:58]: oh! Let me try.

Comment (loganharbour) [reply framework/src/variables/MooseVariableFE.C]: I think that older compilers may not be smart enough to try to grab a member variable in a template like that.

**Review by loganharbour — APPROVED** [review]
Nice! Worked like a charm.

**Comment by YaqiWang** [comment]
Thanks @loganharbour

**Pull request merged**